File: 775-n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var children: [Node]
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.children = []
 *     }
 * }
 */

class Solution {
    func preorder(_ root: Node?) -> [Int] {
        guard let root else { return [] }

        var result = [Int]()
        result.append(root.val)

        root.children.forEach {
            result.append(contentsOf: preorder($0))
        }
        
        return result
    }
}

File: 775-n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.kt
/**
 * Definition for a Node.
 * class Node(var `val`: Int) {
 *     var children: List<Node?> = listOf()
 * }
 */

class Solution {
    fun preorder(root: Node?): List<Int> {
        if (root == null) return emptyList()
        val result = mutableListOf<Int>()
        result.add(root.`val`)
        for (child in root.children) {
            result.addAll(preorder(child))
        }
        return result
    }
}